### PR TITLE
[Streaming Generator] Make it compatible with wait

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR makes the streaming generator compatible with `ray.wait`.

The semantic is as follows;

```python3
def f():
    for _ in range(3):
        yield 1
generator = f.options(num_returns="streaming").remote()
# The generator will be in ready if the next reference is available. Otherwise it is in unready.
# This should work with all other options from ray.wait (including fetch_local=True/False)
ready, unready = ray.wait([generator])

# if the generator's next ref is not ready in 0.1 second, it will be in unready.
# otherwise, it is in ready
ready, unready = ray.wait([generator], timeout=0.1)

# If the generator's next ref is available, it is considered as 1 return
# In this case, this will return if both generator and ref is ready.
ready, unready = ray.wait([generator, ref], num_returns=2)

# if the generator's next ref is available, it will fetch the object to the local node
ready, unready = ray.wait([generator, ref], fetch_local=True)

```

From the previous PR https://github.com/ray-project/ray/pull/36070, we are now able to peek the object reference, and the peeked object is guaranteed to be resolved. We can always peek the next object from the generator and wait on that reference to make the generator compatible to ray.wait.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xieus** [comment]
@jjyao and @cadedaniel could you help review this PR with priority? Thanks.

File: python/ray/tests/test_streaming_generator.py [hunk 1]
```diff
@@ -799,18 +812,170 @@ async def main():
...
+        for i in range(2):
+            time.sleep(sleep_time)
+            yield i
+
+    @ray.remote
+    def g(sleep_time):
+        time.sleep(sleep_time)
+        return 10
+
+    gen = f.options(num_returns="streaming").remote(1)
```
Comment (cadedaniel) [inline python/ray/tests/test_streaming_generator.py:832]: question: how does streaming work when a task both yields and returns values?

File: python/ray/tests/test_streaming_generator.py [hunk 2]
```diff
@@ -799,18 +812,170 @@ async def main():
...
+    assert len(r) == 1
+    assert ray.get(next(r[0])) == 0
+    assert len(ur) == 0
+
+    s = time.time()
+    r, ur = ray.wait([gen], num_returns=1)
+    print(time.time() - s)
+    assert len(r) == 1
+    assert ray.get(next(r[0])) == 1
+    assert len(ur) == 0
```
Comment (cadedaniel) [inline python/ray/tests/test_streaming_generator.py:849]: nit

```suggestion
    for expected_rval in [0, 1]:
        s = time.time()
        r, ur = ray.wait([gen], num_returns=1)
        print(time.time() - s)
        assert len(r) == 1
        assert ray.get(next(r[0])) == expected_rval
        assert len(ur) == 0
```

File: python/ray/tests/test_streaming_generator.py [hunk 3]
```diff
@@ -799,18 +812,170 @@ async def main():
...
+
+    """
+    Test num_returns
+    """
+    gen = f.options(num_returns="streaming").remote(1)
+    ref = g.remote(1)
+    ready, unready = ray.wait([ref, gen], num_returns=2)
+    assert len(ready) == 2
+    assert len(unready) == 0
+
```
Comment (cadedaniel) [inline python/ray/tests/test_streaming_generator.py:914]: are these edge cases important:

* should have a test for `fetch_local=True` where object is larger than 100kb (through object store)? (will go through different path in core worker AFAIK)
* should have a test for `fetch_local=False`?
* is it important to test `await asyncio.wait(object_refs)` ?
* [streaming] does ray gracefully handle a generator raising an exception between yields?

File: python/ray/_private/worker.py [hunk 4]
```diff
@@ -2747,29 +2748,27 @@ def wait(
...
 
-    if isinstance(object_refs, ObjectRef):
+    if isinstance(object_refs, ObjectRef) or isinstance(
+        object_refs, StreamingObjectRefGenerator
+    ):
         raise TypeError(
-            "wait() expected a list of ray.ObjectRef, got a single ray.ObjectRef"
+            "wait() expected a list of ray.ObjectRef or StreamingObjectRefGenerator"
+            ", got a single ray.ObjectRef or StreamingObjectRefGenerator "
+            f"{object_refs}"
```
Comment (cadedaniel) [inline python/ray/_private/worker.py:2757]: nit: `s/StreamingObjectRefGenerator/ray.StreamingObjectRefGenerator/g` or `s/ray.ObjectRef/ObjectRef/g`

File: python/ray/_raylet.pyx [hunk 5]
```diff
@@ -3026,13 +3027,30 @@ cdef class CoreWorker:
...
 
-    def wait(self, object_refs, int num_returns, int64_t timeout_ms,
+    def wait(self, object_refs_or_generator, int num_returns, int64_t timeout_ms,
              TaskID current_task_id, c_bool fetch_local):
         cdef:
             c_vector[CObjectID] wait_ids
             c_vector[c_bool] results
             CTaskID c_task_id = current_task_id.native()
 
+        object_refs = []
```
Comment (cadedaniel) [inline python/ray/_raylet.pyx:3037]: nit: consider pre-allocating array size to avoid high amortized O(1) append? we may see perf regression in ray.wait release test, it waits 10k object refs in a loop IIRC

```python
object_refs = [None] * len(object_refs_or_generator)
for i, ref_or_generator in enumerate(object_refs_or_generator):
  ...
  object_refs[i] = ...

```

fine to ignore, don't want to prematurely optimize

**Review by cadedaniel — APPROVED** [review]

File: python/ray/_raylet.pyx [hunk 6]
```diff
@@ -3026,13 +3027,30 @@ cdef class CoreWorker:
 
         return c_object_id.Binary()
 
-    def wait(self, object_refs, int num_returns, int64_t timeout_ms,
+    def wait(self, object_refs_or_generators, int num_returns, int64_t timeout_ms,
```
Comment (stephanie-wang) [inline python/ray/_raylet.pyx:3030]: This is missing implementation for the fetch_local flag.

File: python/ray/tests/test_streaming_generator.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/tests/test_streaming_generator.py:832]: I think it will be the same as normal Python - `return` just raises StopIteration.

File: python/ray/tests/test_streaming_generator.py (hunk 3 above)
Comment (stephanie-wang) [reply python/ray/tests/test_streaming_generator.py:908]: These are good test cases to add, but let's do it in a follow-up PR. I'll add these as a TODO for @rkooo567.

For the last one, it will return the exception as the last object in the generator. I think there are already tests for that in this file.

Comment (rkooo567) [reply python/ray/tests/test_streaming_generator.py:908]: > should have a test for fetch_local=True where object is larger than 100kb (through object store)? (will go through different path in core worker AFAIK)

Hmm, I originally thought it is just re-testing what ray.wait should have already tested. I guess there's nothing to lose by adding a simple test...

> is it important to test await asyncio.wait(object_refs) ?

This one seems unnecessary (should have already tests somewhere else)

File: python/ray/_raylet.pyx [hunk 7]
```diff
@@ -234,8 +234,15 @@ class StreamingObjectRefGenerator:
         self._generator_task_exception = None
         # Ray's worker class. ray._private.worker.global_worker
         self.worker = worker
+        self.worker.check_connected()
         assert hasattr(worker, "core_worker")
 
+    def get_next_ref(self) -> ObjectRef:
+        self.worker.check_connected()
+        core_worker = self.worker.core_worker
+        return core_worker.peek_object_ref_stream(
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/_raylet.pyx:244]: it is in the previous PR

File: python/ray/tests/test_streaming_generator.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/tests/test_streaming_generator.py:788]: Added a test
